### PR TITLE
Fix #290 Query DI Part 1 - CompiledQueryCache and Database classes

### LINKED ISSUE #290: Query: DI enable
...when architecture stabilizes

### PR DESCRIPTION
**Pull request by mikary**
Move compilation code out of cache to allow for scoped injection
Decouple Database classes from query compilation, eliminates need for provider-
specific relational Database classes

@maumar @anpete

### CODE REVIEW

File: src/EntityFramework.Core/EntityFramework.Core.csproj [hunk 1]
```diff
@@ -217,6 +217,14 @@
     <Compile Include="Query\Annotations\QueryAnnotation.cs" />
     <Compile Include="Query\Annotations\IncludeQueryAnnotation.cs" />
     <Compile Include="EF.cs" />
+    <Compile Include="Query\Compiler\CompiledQuery.cs" />
```
Comment (anpete) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: Please don't introduce new namespaces - we have already done API review here.

File: src/EntityFramework.Core/EntityFramework.Core.csproj [hunk 2]
```diff
@@ -224,17 +232,24 @@
     <Compile Include="Query\Expressions\PropertyEvaluationPreventingExpression.cs" />
     <Compile Include="Query\ExpressionVisitors\ExpressionStringBuilder.cs" />
     <Compile Include="Query\ExpressionVisitors\ExpressionVisitorBase.cs" />
-    <Compile Include="Query\ExpressionVisitors\FunctionEvaluationDisablingVisitor.cs" />
+    <Compile Include="Query\ExpressionVisitors\NullEvaluatableExpressionFilter.cs" />
+    <Compile Include="Query\IQueryCompilationContextFactory.cs" />
+    <Compile Include="Query\Preprocessor\Internal\FunctionEvaluationDisablingVisitor.cs" />
```
Comment (anpete) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: New namespace.

Comment (anpete) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: I don't think we have to pull out all nested classes. Some of them are implementation detail.

File: src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -104,19 +104,34 @@ public static class EntityFrameworkServiceCollectionExtensions
...
+        private static IServiceCollection AddQuery(this IServiceCollection serviceCollection)
+        {
+            return serviceCollection
+                .AddSingleton<IMemoryCache, MemoryCache>()
+                .AddSingleton<ICompiledQueryCache, CompiledQueryCache>()
+                .AddScoped<IEntityQueryProvider, EntityQueryProvider>()
+                .AddScoped<IQueryExecutor, QueryExecutor>()
+                .AddScoped<IQueryPreprocessor, QueryPreprocessor>()
+                .AddScoped<QueryCompiler>()
+                .AddScoped<CompiledQueryCacheKeyGenerator>()
```
Comment (anpete) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: Shouldn't this be registered via interface? Same with QueryCompiler.

File: src/EntityFramework.Core/Query/Compiler/CompiledQuery.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Data.Entity.Query.Compiler
+{
+    public class CompiledQuery
```
Comment (anpete) [inline src/EntityFramework.Core/Query/Compiler/CompiledQuery.cs]: This class is an implementation detail of the cache.

File: src/EntityFramework.Sqlite/Query/SqliteQueryCompilationContextFactory.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+using Microsoft.Data.Entity.ChangeTracking.Internal;
+using Microsoft.Data.Entity.Metadata;
+using Microsoft.Data.Entity.Metadata.Internal;
+using Microsoft.Data.Entity.Query.Methods;
+using Microsoft.Data.Entity.Storage;
+using Microsoft.Framework.Logging;
+
+namespace Microsoft.Data.Entity.Query
+{
+    public class SqliteQueryCompilationContextFactory : IQueryCompilationContextFactory
```
Comment (anpete) [inline src/EntityFramework.Sqlite/Query/SqliteQueryCompilationContextFactory.cs]: Isn't the point of this change to make this go away?

File: src/EntityFramework.Core/EntityFramework.Core.csproj (hunk 2 above)
Comment (mikary) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: I would rather put them in M.D.E.Query.Internal, they are implementation details but the purpose of the precompiler is to run through a series of transformations and the details shouldn't matter (Trying for more of an open/closed principal approach)

File: src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs (hunk 3 above)
Comment (mikary) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: It's registered as a provider service to allow custom cache key generation and this implementation is simply the default. It's the same with QueryCompiler, though at this point only the InMemory provider needs to override.

File: src/EntityFramework.Sqlite/Query/SqliteQueryCompilationContextFactory.cs (hunk 5 above)
Comment (mikary) [inline src/EntityFramework.Sqlite/Query/SqliteQueryCompilationContextFactory.cs]: It probably will go away in the long run, but this PR is intended to be an incremental change. We can hold off on merging if it would be better to have the full implementation but I didn't want too overwhelming of a PR

File: src/EntityFramework.Core/EntityFramework.Core.csproj (hunk 2 above)
Comment (anpete) [inline src/EntityFramework.Core/EntityFramework.Core.csproj]: That's fine when they are not purely implementation detail.

File: src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs (hunk 3 above)
Comment (anpete) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: I'm referring to lines 127, 128

Comment (mikary) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: There are exceptions required for the `DatabaseProviderServices`, for example the `ValueGeneratorSelector` is registered this way (though in that case both InMemory and Relational have implementations registered so it might not be necessary).

I originally had the Core compiler as an abstract class with InMemory and Relational implementations, but the logic for relational just duplicated what happens in core.

Comment (anpete) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: OK, I guess I must be missing something about how this works. I guess another question is: Is it required to be this way in order to work?

Comment (mikary) [inline src/EntityFramework.Core/EntityFrameworkServiceCollectionExtensions.cs]: Anything else would be going against the pattern for `DatabaseProviderServices` or result in duplicated code (registering as a class in the AddProvider methods or being completely abstract and using the same logic in InMemory and Relational)

File: src/EntityFramework.Core/Query/Compiler/CompiledQueryCacheKeyGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq.Expressions;
+using JetBrains.Annotations;
```
Comment (maumar) [inline src/EntityFramework.Core/Query/Compiler/CompiledQueryCacheKeyGenerator.cs]: nit - move JetBrains using to the end

File: src/EntityFramework.Core/Query/QueryCompiler.cs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+                    return _database.CompileAsyncQuery<TResult>(queryModel);
+                });
+        }
+
+        private Func<QueryContext, TResult> MapQueryExecutor<TResult>(QueryModel queryModel, Type resultItemType)
+        {
+            if (resultItemType == typeof(TResult))
+            {
+                var func = _database.CompileQuery<TResult>(queryModel);
```
Comment (maumar) [inline src/EntityFramework.Core/Query/QueryCompiler.cs]: compiledQuery rather than func?

File: src/EntityFramework.Core/Query/QueryPreprocessor.cs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            {
+                var clonedModel = expression.QueryModel.Clone();
+
+                clonedModel.TransformExpressions(Visit);
+
+                return new SubQueryExpression(clonedModel);
+            }
+        }
+
+        public class NullEvaluatableExpressionFilter : EvaluatableExpressionFilterBase
```
Comment (maumar) [inline src/EntityFramework.Core/Query/QueryPreprocessor.cs]: private?

File: src/EntityFramework.Core/Query/QueryPreprocessor.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                return base.VisitMethodCall(methodCallExpression);
+            }
+        }
+
+        private class FunctionEvaluationDisablingVisitor : ExpressionVisitorBase
+        {
+            public static readonly MethodInfo DbContextSetMethodInfo
+                = typeof(DbContext).GetTypeInfo().GetDeclaredMethod("Set");
+
+            private static readonly MethodInfo[] _nonDeterministicMethodInfos =
```
Comment (maumar) [inline src/EntityFramework.Core/Query/QueryPreprocessor.cs]: fix formatting

File: src/EntityFramework.Core/Query/QueryPreprocessor.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            var partialEvaluationInfo = EvaluatableTreeFindingExpressionVisitor
+                .Analyze(functionEvaluationDisabledExpression, new NullEvaluatableExpressionFilter());
+
+            var parameterExtractedExpression = new ParameterExtractingExpressionVisitor(partialEvaluationInfo, queryContext)
+                .Visit(functionEvaluationDisabledExpression);
+
+            return parameterExtractedExpression;
+        }
+
+        private class NullEvaluatableExpressionFilter : EvaluatableExpressionFilterBase
```
Comment (anpete) [inline src/EntityFramework.Core/Query/QueryPreprocessor.cs]: Don't duplicate this..

File: src/EntityFramework.Core/Query/QueryPreprocessor.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+using System.Linq.Expressions;
+using JetBrains.Annotations;
+using Microsoft.Data.Entity.Query.ExpressionVisitors;
+using Microsoft.Data.Entity.Utilities;
+using Remotion.Linq.Parsing.ExpressionVisitors.TreeEvaluation;
+
+namespace Microsoft.Data.Entity.Query
+{
+    public class QueryPreprocessor : IQueryPreprocessor
```
Comment (anpete) [inline src/EntityFramework.Core/Query/QueryPreprocessor.cs]: This new abstraction seems unrelated to the DI change.

**Comment by mikary** [comment]
Updated to remove unrelated abstractions

**Comment by anpete** [comment]
:shipit:
